doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,530 @@
+# SONiC GNMI Server Interface Design
+
+# High Level Design Document
+
+#### Rev 0.1
+
+# Table of Contents
+- [Table of Contents](#table-of-contents)
+- [List of Tables](#list-of-tables)
+- [Revision](#revision)
+- [About this Document](#about-this-document)
+- [Scope](#scope)
+- [Definition/Abbreviation](#definitionabbreviation)
+- [1 Feature Overview](#1-feature-overview)
+  * [1.1 Requirements](#11-requirements)
+  * [1.2 Design Overview](#12-design-overview)
+- [2 Error Handling](#2-error-handling)
+- [3 Serviceability and Debug](#3-serviceability-and-debug)
+- [4 Unit Tests](#4-unit-tests)
+  * [4.1 Unit Tests for Capabilities RPC](#41-unit-tests-for-capabilities-rpc)
+  * [4.2 Unit Tests for Get RPC](#42-unit-tests-for-get-rpc)
+  * [4.3 Unit Tests for Set RPC](#43-unit-tests-for-set-rpc)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+[Table 2: Mapping Operations](#table-2-mapping-operations)
+
+[Table 3: Translate Example](#table-3-translate-example)
+
+# Revision
+
+| Rev | Date        | Author             | Change Description  |
+|:---:|:-----------:|:------------------:|---------------------|
+| 0.1  | 05/05/2022 | Gang Lv    | Initial version     |
+
+# About this Document
+This document provides a detailed description on the strategy to implement the SONiC GNMI Server Interface.
+
+# Scope
+This document describes the high level design of SONiC GNMI Server Interface. 
+
+This document provides minor implementation details about the proposed solutions.
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+| **Term** | **Meaning**                |
+| -------- | -------------------------- |
+| gRPC     | A modern open-source high performance Remote Procedure Call (RPC) framework that can run in any environment |
+| gNMI     | gRPC Network Management Interface, used to retrieve or manipulate the state of a device via telemetry or configuration data |
+| gNOI     | gRPC Network Operations Interface |
+| JSON     | JavaScript Object Notation |
+| Yang     | Yet Another Next Generation, it is a modular language representing data structures in an XML tree format |
+| Generic Config Updater | SONiC configuration generic update and rollback feature |
+| ConfigDB | Configuration Database |
+| ApplDB | Application Database |
+
+# 1 Feature Overview
+
+SONiC should provide a gNMI/gNOI server interface where the client can communicate with SONiC (server) through the interface using their respective gRPC libraries based on ConfigDB/ApplDB/StateDB/CountersDB schema or SONiC Yang schema.
```

Comment:
Looks like you are opening up gNMI to write directly to APPDB (any DB) without any application - this may result in data consistency issues? for security concerns, someone bypasses some checks and directly operates the DB. what is the plan to mitigate?

Reply:
gNMI can only update table in ApplDB with Yang models.

Reply:
@madhupalu The requirement comes from DASH project. The gNMI server will only write to limited tables in ApplDB, and these tables must have Yang model defined. In this use case, gNMI play the same role as a RestAPI server.

Reply:
sounds good.